### ISSUE TITLE
remote debug for bashdb.

### ISSUE THREAD

**Issue by kenjo** [realgud/realgud#166]
while I managed to make it start a remote session with
`M-x  realgud:bashdb RET`
then setting the program to run as. 
`ssh -t root@192.168.1.1 bashdb BashScriptToDebug`

it fails when bashdb prints out the sourcefile and line number. There needs to be some way to convert the target local path to a remote path that emacs understands. 

ex the bashdb printout 
`(/root/testing.sh:13):`

needs to be prefixed with a TRAMP path like `/root:192.168.1.1:` then emacs can load the file using the normal `find-file` command

**Comment by rocky** [comment]
`realgud-file-remap` is a hash for that purpose. The docstring says: 
```
  "How to remap files we otherwise can't find in the
  filesystem. The hash key is the file string we saw, and the
  value is associated filesystem string presumably in the
  filesystem"
```

And it should be the case that when realgud can't find the file it needs it prompts for a name that goes into this hash. So I'm guessing that it is finding the file on the local file system and using that. 

Another way to approach this is to write a file remap function and attach that. The jdb debugger does this so see function  `realgud:jdb-find-file`. 

I know this is a little sketchy, but right now this is about all the time I can spend on right now.

**Comment by kenjo** [comment]
Yes that was a bit over my head at the moment. One more thing I noticed is that I have to start emacs in / to be able to start the debugger. otherwise when the command `realgud:bashdb` ask for the debugger to run and I input `ssh -t root@192.168.1.1 bashdb BashScriptToDebug` it adds in the current directory so the commands is `ssh -t  /current/directory/root@192.168.1.1 bashdb BashScriptToDebug` and that do not work. 

Problem is that I don't really know how to develop elisp so I have a hard time doing the changes. But it feels like there is only a few  small changes that needs to be done to get remote debug to work with bash.

**Comment by YorkZ** [comment]
First of all, `bashdb` and `realgud` are so cool and useful, thank you so much.
I'm also very interested in remote debugging bash script in local Emacs, but I didn't see the variable `realgud-file-remap`, but only `realgud:trepan-file-remap`. Can you please let me know how I can make remote debugging for for `bashdb`? Thanks in advance.

**Comment by rocky** [comment]
Thanks for the kind words. Sorry for the delayed reply. Open source should be about others participating and helping. Too often though it is just me.  That means that unless you are say sponsoring the project, it may slip by for a while. 

As to your question. I just add a remap variable for bashdb call `realgud:bashdb-file-remap`.

The way it should work is that when you are on a remote machine and a file isn't found bash should prompt for the filename. 

```Unable to locate /tmp/foo5.sh
Blacklist it for location tracking? (y or n)
```
If you type "n" then you should see:

```
Find this error in (default "foo5.sh"): <some path>
```

Enter the name of the local file that this appears in. Let's say it is `/tmp/foo.sh` so you enter that. 

At this point there is a nasty intermittent bug in realgud that causes minibuffer input not to allow any input. When that happens I am entering Ctr-G to kill this and but when I can enter input, then the file is remapped.

If you look at Debugger Info you should see something like this: 


```
* General Information (realgud-cmdbuf-info)
  - Debugger name       ::	bashdb
  - Command-line args   ::	null
  - Starting directory  ::	nil
  - Current source-code path  ::	[[/tmp/foo.sh]]
  - Selected window should contain source? :: nil
  - Last input end      ::	913695
  - Source should go into short-key mode? :: t
  - In debugger?        ::	t
  - Ignore file regexps ::	( /tmp/foo3.sh)
  - remapped file names ::	((/tmp/foo5.sh . /tmp/foo.sh))
```

Under "Ignore file regexps" are the situations where you decide not to remap a file. Here is it `/tmp/foo3.sh`.
Under `remapped file names` should be a hash of the file paths you want to remap.

You can set this in advance with Elisp:

```
(setf (gethash "/tmp/foo.sh" realgud:bashdb-file-remap) "/tmp/foo5.sh")

```
